File: 0016-3sum-closest/solution.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) 
    {
        const int n = nums.size();
        int diff = INT_MAX;
        
        sort(nums.begin(), nums.end());
        int sum;
        for(int i = 0; i < n; i++)
        {
            int lo = i + 1, hi = n - 1;
            while(lo < hi)
            {
                sum = nums[i] + nums[lo] + nums[hi];
                
                if(abs(target - sum) < abs(diff)) diff = target - sum;                
                if(sum < target) lo++;
                
                else hi--;
            }
            if(diff == 0) break;
        }
        
        return target - diff;
    }
};

File: 0016-3sum-closest/solution.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        diff = float('inf')
        
        
        for i in range(n):
            lo, hi = i+1, n - 1
            while lo < hi:
                sum_ = nums[i] + nums[lo] + nums[hi]
                
                if abs(target - sum_) < abs(diff):
                    diff = target - sum_
                if sum_ < target:
                    lo += 1
                else:
                    hi -= 1
            if diff == 0:
                break
        return target - diff
